src/devices/Tlc1543/Tlc1543.cs

```diff
@@ -0,0 +1,221 @@
+            _chipSelect = chipSelect;
+            _dataOut = dataOut;
+            _endOfConversion = endOfConversion;
+            _inputOutputClock = inputOutputClock;
+
+            _controller.OpenPin(_address, PinMode.Output);
+            _controller.OpenPin(_chipSelect, PinMode.Output);
+            _controller.OpenPin(_dataOut, PinMode.InputPullUp);
+            _controller.OpenPin(_inputOutputClock, PinMode.Output);
+            if (_endOfConversion != -1)
+            {
+                _controller.OpenPin(_endOfConversion, PinMode.InputPullUp);
+            }
+        }
+
+        #endregion
+
+        #region Methods
+
+        /// <summary>
+        /// Check if channelNumber is valid
+        /// </summary>
+        /// <param name="channelNumber">Channel Address</param>
+        /// <returns></returns>
+        protected static bool IsValidChannel(int channelNumber)
+        {
+            if (channelNumber < 0 || channelNumber > 13)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Reads sensor value
+        /// <remarks>
+        /// <br>1 cycle used - gets data on second call of this method</br>
+        /// <br>First cycle: Ask for value on the channel</br>
+        /// </remarks>
+        /// </summary>
+        /// <param name="channelNumber">Channel to read</param>
+        /// <returns>10 bit value corresponding to relative voltage level on channel</returns>
+        public int ReadChannel(int channelNumber)
+        {
+            if (!IsValidChannel(channelNumber))
+            {
+                throw new ArgumentOutOfRangeException("channelNumber", "Out of range. \nMust be non-negative and less than 14. \nRefer to table 2 and 3 in TLC1543 documentation for more information");
+            }
+
+            int value = 0;
+            _controller.Write(_chipSelect, 0);
+            for (int i = 0; i < 10; i++)
+            {
+                if (i < 4)
+                {
+                    // send 4-bit Address
+                    if (((channelNumber >> (3 - i)) & 0x01) == 1)
+                    {
+                        _controller.Write(_address, 1);
+                    }
+                    else
+                    {
+                        _controller.Write(_address, 0);
+                    }
+                }
+
+                // read 10-bit data
+                value <<= 1;
+                if (_controller.Read(_dataOut) == PinValue.High)
+                {
+                    value |= 0x01;
+                }
+
+                _controller.Write(_inputOutputClock, 1);
+                DelayHelper.DelayMicroseconds(0, false);
+                _controller.Write(_inputOutputClock, 0);
+            }
+
+            _controller.Write(_chipSelect, 1);
+            // Max conversion time (21us) as seen in table on page 10 in TLC1543 documentation
+            DelayHelper.Delay(_conversionTime, false);

```

Comment:
you should not implement SPI, rather you should take SPI in the constructor - for local testing you can use SoftwareSpi which takes GpioController and pin numbers

Reply:
most likely for 10 bit data you can simply ReadWrite and pass zeros for ignored parts